Do metadata sync in a separate background worker.

pykello opened the pull request:
DESCRIPTION: Fix distributed deadlock detection being blocked by metadata sync

Metadata syncing several locks (on pg_dist_node, pg_dist_partition, distributed tables, ...) and can be blocked by other backends. So it shouldn't run in the same process as distributed deadlock detection. Otherwise a metadata sync which is blocked by backends involved in a distributed deadlock can cause the system to get stuck and prevent progress.

To solve it, this PR moves metadata sync to a separate background worker which is spawn from maintenanced.

Fixes #4393.

marcocitus commented on src/backend/distributed/metadata/metadata_sync.c:
```diff
@@ -1539,3 +1545,105 @@ SyncMetadataToNodes(void)
+	worker.bgw_flags =
+		BGWORKER_SHMEM_ACCESS | BGWORKER_BACKEND_DATABASE_CONNECTION;
+	worker.bgw_start_time = BgWorkerStart_ConsistentState;
+	worker.bgw_restart_time = (MetadataSyncRetryInterval + 999) / 1000;
```
what's the +999?

pykello replied:
This is to get the ceiling of division by 1000 instead of its floor. bgw_restart_time is integer seconds while MetadataSyncRetryInterval is in milliseconds.

Alternatively we could:
1. wait in precise milliseconds by implementing a loop inside the background worker's main, which might make the code a bit more complicated.
2. change units of MetadataSyncRetryInterval to seconds

I will try to implement option 2.

marcocitus commented on src/backend/distributed/metadata/metadata_sync.c:
```diff
@@ -1539,3 +1545,105 @@ SyncMetadataToNodes(void)
+
+		/*
+		 * Notification means we had an attempt on synchronization
+		 * without being blocked for pg_dist_node access.
```
Not really due to this PR, but I had some trouble understanding this comment. It could be nice to add something like "We use LISTEN/NOTIFY to wait for metadata syncing in tests".

pykello replied:
done

marcocitus reviewed (commented):
Seems like a nice and simple solution!

pykello commented:
@marcocitus tests are passing now. I still need to address your feedback and:
1. test restart behaviour
2. understand what happens if maintenance daemon aborts. Do we have an orphan background worker or maintenance daemon exit also exits metadata sync?

pykello commented:
@marcocitus I think this is ready to be reviewed.

pykello commented on src/test/regress/spec/isolation_metadata_sync_deadlock.spec:
```diff
@@ -0,0 +1,104 @@
+    SELECT pg_sleep(1);
+}
+
+permutation "enable-deadlock-detection" "reload-conf" "s2-start-session-level-connection" "s1-begin" "s1-update-1" "s2-begin-on-worker" "s2-update-2-on-worker" "s2-truncate-on-worker" "s3-invalidate-metadata-and-resync" "s2-update-1-on-worker" "s1-update-2" "s1-commit" "s2-commit-on-worker" "disable-deadlock-detection" "reload-conf"
```
Add a comment on what this is checking.

pykello commented on src/test/regress/sql/multi_mx_node_metadata.sql:
```diff
@@ -27,6 +27,20 @@ CREATE FUNCTION mark_node_readonly(hostname TEXT, port INTEGER, isreadonly BOOLE
+   counter integer := -1;
+begin
+ while counter != target_count loop
+  SELECT into counter FROM pg_stat_activity WHERE application_name = appname;
```
i'm not sure how this worked, but this should be SELECT count(*) into counter ...

marcocitus replied:
I'm guessing this worked because it reached NULL at some point. Agree it should be count(*).

marcocitus commented on src/test/regress/expected/isolation_metadata_sync_deadlock.out:
```diff
@@ -0,0 +1,86 @@
+run_commands_on_session_level_connection_to_node
+
+
+error in steps s1-update-2 s2-update-1-on-worker: ERROR:  canceling the transaction since it was involved in a distributed deadlock
```
cool, thanks for adding this test

marcocitus commented on src/backend/distributed/utils/maintenanced.c:
```diff
@@ -626,6 +623,11 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 		}
 	}
+
+	if (metadataSyncBgwHandle)
```
do you know what happens if the maintenance daemon exits through an error?

pykello replied:
Instead of doing this, I changed start-up of maintenance daemon to send SIGTERM to metadata sync daemons of crashed maintenance daemons.

Removed this so tests verify that approach.

marcocitus commented on src/backend/distributed/metadata/metadata_sync.c:
```diff
@@ -1527,6 +1533,9 @@ SyncMetadataToNodes(void)
 			{
+				ereport(WARNING, (errmsg("failed to sync metadata to %s:%d",
+										 workerNode->workerName,
+										 workerNode->workerPort)));
```
do we hit this warning in case of a deadlock?

pykello replied:
If the daemon itself is cancelled because of distributed deadlock, we should get  `ERROR: canceling statement due to user request`, which will cause the daemon to `exit(1)` (in elog.c's handlers), and restarted in few seconds.

If another backend which is blocking the metadata_sync is cancelled, `SyncMetadataSnapshotToNode` should be able to succeed as it got all required locks.

This WARNING can happen in situations like connection problems or when workerNode is read-only.

pykello commented:
@marcocitus I fixed termination/cancellation behaviour and added some tests (one more isolation test + few multi-mx tests) to verify the changes I made.

pykello commented on src/backend/distributed/metadata/metadata_sync.c:
```diff
@@ -1539,3 +1556,216 @@ SyncMetadataToNodes(void)
+		 * If backend is cancelled (e.g. bacause of distributed deadlock), exit with
+		 * code 1 so postgres restarts the daemon in few seconds.
+		 */
+		if (QueryCancelPending)
```
This is not really necessary, since `CHECK_FOR_INTERRUPTS` should error out for cancellations.

marcocitus commented on src/backend/distributed/metadata/metadata_sync.c:
```diff
@@ -1539,3 +1556,211 @@ SyncMetadataToNodes(void)
+	worker.bgw_flags =
+		BGWORKER_SHMEM_ACCESS | BGWORKER_BACKEND_DATABASE_CONNECTION;
+	worker.bgw_start_time = BgWorkerStart_ConsistentState;
+	worker.bgw_restart_time = retryIntervalSeconds;
```
could this background worker be duplicated, like we sometimes see when maintenanced hits OOM?

pykello replied:
Yes, it is possible. I fixed a related issue in https://github.com/citusdata/citus/pull/4419/commits/1f508a2e87b44eebcf8695667baaf6603ea73e67. I'll fix it properly by end of Monday.

marcocitus commented on src/backend/distributed/utils/maintenanced.c:
```diff
@@ -450,47 +462,76 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 			{
 				ereport(DEBUG1, (errmsg("could not lock the citus extension, "
 										"skipping metadata sync")));
+				nextTimeout = MetadataSyncRetryInterval;
```
it kind of seems like we could just set it to MetadataSyncRetryInterval above and skip setting it here and in the `else { ... }`

marcocitus commented on src/backend/distributed/metadata/metadata_sync.c:
```diff
@@ -1539,3 +1556,243 @@ SyncMetadataToNodes(void)
+		}
+		else if (CheckCitusVersion(DEBUG1) && CitusHasBeenLoaded())
+		{
+			MetadataSyncResult result = SyncMetadataToNodes();
```
One observation as it relates to distributed deadlocks:

We seem to send `BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 0, '2000-01-01 01:00:00+01');` to workers when doing metadata syncing. I wonder if we need a `UseCoordinatedTransaction()` here.

metdos commented:
Shall we also add a description to this PR? I believe we also plan to backport this, @pykello, @marcocitus?

marcocitus commented on src/backend/distributed/metadata/metadata_sync.c:
```diff
@@ -1539,3 +1556,244 @@ SyncMetadataToNodes(void)
+
+	bool syncedAllNodes = false;
+
+	while (!syncedAllNodes)
```
Given this loop, it would be helpful is to only sync one node at a time (i.e. one node per transaction). That way we mostly avoid deadlocks that involve pg_dist_node records.

Not critical point, since this PR is mainly about detecting the deadlock, but having fewer of them would be good as well.

pykello replied:
I'm keeping this for a separate PR.

marcocitus commented on src/backend/distributed/transaction/transaction_management.c:
```diff
@@ -262,6 +266,11 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
 
+			if (MetadataSyncOnCommit)
+			{
+				TriggerMetadataSync(MyDatabaseId);
```
I guess this is still a bit of a race condition, but the window is short and the cost of the metadata sync still starting too early, so probably ok.

One thing we could consider is to put this in post-commit, because at that point the record is actually visible.

marcocitus replied:
would be good to add a comment on why this is here (and needs to be in XACT_EVENT_COMMIT, and not earlier).
